subworkflows/local/metagenomics_profiling.nf

```diff
@@ -0,0 +1,160 @@
+                    // re-extract meta from file names, use filename without rma to
+                    // ensure we keep paired-end information in downstream filenames
+                    // when no pair-merging
+                    def meta_new = meta.clone()
+                    meta_new['db_name'] = meta.id
+                    meta_new['id'] = rma.baseName
+                [ meta_new, rma ]
+            }
+
+        ch_versions                 = ch_versions.mix( MALT_RUN.out.versions.first() )
+        ch_raw_classifications      = ch_raw_classifications.mix( ch_maltrun_for_megan )
+        ch_multiqc_files            = ch_multiqc_files.mix( MALT_RUN.out.log )
+        ch_postprocessing_input     = ch_postprocessing_input.mix( ch_maltrun_for_megan )
+    }
+
+    else if ( params.metagenomics_profiling_tool == 'metaphlan' ) {
+
+        reads = reads.combine(database)
+        metaphlan_reads = reads.map{ meta, reads, database -> [meta, reads] }
+        metaphlan_db = reads.map{ meta, reads, database -> [database] }
+
+        METAPHLAN_METAPHLAN ( metaphlan_reads , metaphlan_db )
+        ch_versions             = ch_versions.mix( METAPHLAN_METAPHLAN.out.versions.first() )
+        ch_raw_profiles         = ch_raw_profiles.mix( METAPHLAN_METAPHLAN.out.profile )
+        ch_postprocessing_input = ch_postprocessing_input.mix( METAPHLAN_METAPHLAN.out.profile )
+
+    }
+
+    if ( params.metagenomics_profiling_tool == 'krakenuniq' ) {
+        // run kraken uniq per sample, to preserve the meta-data
+
+        reads = reads.combine(database)
+        krakenuniq_reads = reads.map{ meta, reads, database -> [meta, reads] }
+        krakenuniq_db = reads.map{ meta, reads, database -> [database] }
+
+        KRAKENUNIQ_PRELOADEDKRAKENUNIQ (
+            krakenuniq_reads,
+            krakenuniq_db,
+            params.metagenomics_krakenuniq_ram_chunk_size,
+            params.metagenomics_kraken_save_reads,
+            true,
+            params.metagenomics_kraken_save_read_classifications
+        )
+
+        ch_multiqc_files           = ch_multiqc_files.mix( KRAKENUNIQ_PRELOADEDKRAKENUNIQ.out.report )
+        ch_versions                = ch_versions.mix( KRAKENUNIQ_PRELOADEDKRAKENUNIQ.out.versions.first() )
+        ch_raw_classifications     = ch_raw_classifications.mix( KRAKENUNIQ_PRELOADEDKRAKENUNIQ.out.classified_assignment )
+        ch_raw_profiles            = ch_raw_profiles.mix( KRAKENUNIQ_PRELOADEDKRAKENUNIQ.out.report )
+        ch_multiqc_files           = ch_multiqc_files.mix( KRAKENUNIQ_PRELOADEDKRAKENUNIQ.out.report )
+        ch_postprocessing_input    = ch_postprocessing_input.mix( KRAKENUNIQ_PRELOADEDKRAKENUNIQ.out.report )
+    }
+
+    else if ( params.metagenomics_profiling_tool == 'kraken2' ) {
+        // run kraken2 per sample
+
+        reads = reads.combine(database)
+        kraken2_reads = reads.map{meta, reads, database -> [meta, reads]}
+        kraken2_db = reads.map{meta, reads, database -> [database]}
```

Comment:
Again not following this

Reply:
database is a queue-channel at this point. But if we run kraken only once the combine becomes unnecessary.